### PR TITLE
DNSBL: never restart Unbound for a slow zero-downtime swap; firewall: detach the filter reload

### LINKED ISSUE #3200: DNSBL datapath swap always misses the 30s wait and restarts Unbound
## Objective

A DNSBL **data-only** reload (feed/cron, Alerts/Reports whitelist, #51 unlock) must apply via the ADR-10 zero-downtime swap and **must not stop Unbound**. Today every such reload on a large python DNSBL hits `pfb_unbound_py_wait_applied()`'s 30s budget, logs `not confirmed in time -- falling back to Unbound restart`, and `pfb_stop_start_unbound()` drops DNS.

This is **one** issue for every `$datapath=TRUE` caller of `pfb_reload_unbound()`. Alerts whitelist is not special. It is **not** #3125 (package upgrade restart when `pfb_unbound.py` bytes changed — that path is working as designed).

## Symptom (live, confirmed 2026-09-04, 10.1.41.3 / pfSense Plus 25.11, DNSBL python, ~1.1M loaded entries)

`sudo grep` of `/var/log/pfblockerng/pfblockerng.log`:

| Pattern | Count |
| --- | --- |
| `[ zero-downtime swap ]` | 86 |
| `falling back to Unbound restart` | 86 |
| `zero-downtime swap ] completed` | **0** |
| `RAM-constrained` | 0 |

Hourly cron (~:15) every hour:

```text
07:15:31 Reloading Unbound Resolver (DNSBL python) [ zero-downtime swap ] not confirmed in time -- falling back to Unbound restart
07:16:01 Reloading Unbound Resolver (DNSBL python)
07:16:01 Stopping Unbound Resolver
```

Alerts/Reports whitelist (same function, same fail-safe):

```text
11:40:27 Reloading Unbound Resolver (DNSBL python) [ zero-downtime swap ]
11:40:55 ... not confirmed in time -- falling back to Unbound restart
11:40:57 Stopping Unbound Resolver
11:41:06 Starting Unbound Resolver..
11:41:37 Unbound control [ status ] TIMED OUT after 30s
```

After the fallback start, `unbound-control status` often also times out at 30s (cold-start of 1.1M entries). Treat that as a **symptom of the same slowness**, not a second root cause, unless a probe shows the swap *would* have applied if the wait were longer vs the watcher never applying.

Nightly install at 14:26 (`DNSBL Unbound python module changed -- Resolver restart required`) is **out of scope** — #3125, module fingerprint, `pfb_unbound.py` changed by #3193.

## Cause (code)

`pfblockerng_alerts.php:1128` and `pfb_update_unbound()` both call `pfb_reload_unbound(..., $datapath=TRUE)`. Fast path (`pfblockerng.inc:12146`): flip sentinel, `pfb_unbound_py_wait_applied($gen)` default **30s** (`:10492`). Timeout logs case 2 ` not confirmed in time -- falling back to Unbound restart` (leading space, no newline — also pollutes `error.log` by concatenating onto the previous line; do not mix a logging fix unless it is free) then `pfb_stop_start_unbound()`.

The watcher rebuilds the matcher from the patched manifest; it does not incrementally patch one whitelist line into live `whiteDB`. On this box that rebuild does not confirm within 30s.

Widening the 30s budget alone is **not** the accepted fix if the box regularly exceeds it and the fallback still restarts (testing.md: duration is not an assertion; a larger constant is not a flake fix).

## Required reading

- `.agents/policy/testing.md` (test-first; duration never an assertion)
- `.agents/policy/coding.md` + `.agents/context/lang-php.md` + `.agents/context/lang-python.md` as touched
- `.agents/policy/git.md` (`work-branch.sh --worktree --claim`)
- `.agents/policy/landing.md` / `.agents/policy/workflow.md`
- `docs/misc/architecture-notes.md` DNSBL/ABP pipeline (ADR-10 zero-downtime)
- `docs/misc/external-process-waits.md`
- `src/usr/local/pkg/pfblockerng/pfblockerng.inc:10486` `pfb_unbound_py_wait_applied`
- `src/usr/local/pkg/pfblockerng/pfblockerng.inc:12129` `pfb_reload_unbound` datapath / fallback
- `src/usr/local/pkg/pfblockerng/pfblockerng.inc:12418` `pfb_update_unbound` `$datapath = !$pfbpython`
- `src/usr/local/www/pfblockerng/pfblockerng_alerts.php:1120` whitelist add (and sibling datapath callers — grep, do not trust this list)
- Python reload watcher / applied marker write in `pfb_unbound.py` (name from codegraph, not memory)
- `tests/smoke/test_smoke_matrix.py` `test_dnsbl_sentinel_flip_failure_clears_sync_marker` (pins restart fallback on sentinel failure — do not break `.sync` cleanup)
- Issue #3125 / PR #3126 (module fingerprint restart — **do not change**)
- Issue #3198 / PR #3199 (Reports save gate — **do not change**)

## Constraints

- Dedicated worktree off current `origin/devel`: `scripts/agent/work-branch.sh issue <NN> --worktree --claim`.
- Test-first: reproduction RED on untouched production, freeze `git hash-object`, production edit, GREEN unchanged.
- **Do not restart Unbound** when a datapath swap does not confirm in time. Queries keep using the old snapshot until the new generation applies (ADR-10: queries NEVER stop during the wait). Returning the UI before live is acceptable; stopping the resolver is not.
- Discriminate **before** choosing a mechanism (debug skill): is `wait_applied` FALSE because the rebuild is slow, the watcher is stuck, or a Python RAM re-check declined? Probe with competing hypotheses. Inventing an incremental `whiteDB` patcher the packet did not name is an ESCALATE / DONE-WITH-DEVIATION, not silent.
- Do not change #3125 module-fingerprint restart (`$pfbpython` / `$datapath=FALSE`).
- Do not change Reports save-gate (#3198).
- Do not "fix" this only in `addwhitelistdom` — cron/feed is the same fail-safe and is the hourly outage.
- Do not use a larger sleep/deadline as the assertion that the swap worked.
- Live box `10.1.41.3` may be re-read (planner now has `sudo -n`); do not weaken DNSBL or rewrite config. Quote command + tail for any live claim.
- Grok spawn-depth 1: after the PR is open, checkpoint with PR URL + head SHA and **stop**. Orchestrator spawns landing.md legs.
- Canonical mandates stay in force.

## Coverage matrix

Enumerate from source (do not trust memory):

```text
git grep -n 'pfb_reload_unbound(' origin/devel -- src/
```

Every `$datapath=TRUE` site must keep Unbound up when the swap is in flight. `$datapath=FALSE` / `$pfbpython=TRUE` (ini, mounts, python module fingerprint) still restarts. Sentinel-flip **failure** (not timeout) may still fail-safe as today (`test_dnsbl_sentinel_flip_failure_clears_sync_marker`); say so if that branch stays.

PHPUnit/unit rows that must exist (names may vary):

1. Datapath + `wait_applied` FALSE (timeout) does **not** call `pfb_stop_start_unbound` / does not log the restart fallback as the chosen path — pin the new behaviour with a fake wait.
2. Datapath + `wait_applied` TRUE still returns TRUE and does not restart (happy swap).
3. Datapath + sentinel flip FALSE still takes the documented fail-safe (if still restart) and still clears `.sync`.
4. Non-datapath / `$pfbpython` restart path unchanged.

If PHPUnit cannot stub `is_process_running` / RAM gate, say how the existing UnboundRestartSeam / publish tests do it and reuse that seam. Do not add a 30-second sleep test.

## Hostile-input rows

No new parser. Marker file contents: missing, empty, non-numeric, generation behind, generation equal, generation ahead — already `pfb_unbound_py_wait_marker`. Do not add a second parser.

## Risk triggers

**Live-appliance behaviour.** Extra verification: after the fix, a datapath reload on a large DNSBL (smoke VM with a non-trivial list, or a quoted live grep on 10.1.41.3 if the owner is using that box) must log `[ zero-downtime swap ]` **completed** (or an equivalent "still serving old snapshot, applied later" line) and must **not** log `Stopping Unbound Resolver` for that reload. If a smoke seat is unavailable after a real attempt, checkpoint — do not skip silently.

## Verification

- WHEN datapath reload's `wait_applied` times out THEN Unbound process is not stopped
- WHEN the same reload's applied marker later reaches `$gen` THEN queries see the new whitelist/data without an intervening stop/start
- WHEN reproduction tests run on untouched `origin/devel` THEN they FAIL because timeout still falls through to `pfb_stop_start_unbound`
- WHEN those frozen tests re-run after the production edit THEN they PASS with byte-identical files
- WHEN `scripts/agent/run-gates.sh --diff origin/devel` THEN green for touched types
- WHEN PR is opened THEN `Fixes #<this>` and executed red→green pasted
- WHEN landing.md four legs finish (orchestrator-spawned) and exact-head CI is green THEN land

## Escalation

STOP and checkpoint + `needs-info` / `ready-for-human` if:

- Live probe shows the watcher **never** applies even after minutes (stuck, not slow) and the packet's "keep old snapshot" would leave whitelist/feed changes dead
- A named mechanism this packet never listed is required (new config key, Unbound pythonmod API, incremental whiteDB mutator) — return that as the fork, do not silently invent it
- #3125 fingerprint path would be broken by the change

### PR DESCRIPTION
**Pull request by andrebrait**
Fixes #3200

Two related behavior changes, per the owner directive on the ticket: reload waits that treat elapsed time as failure are removed or made non-fatal — the only fail-safe triggers are genuine errors or definite breakage.

## Commit 1 — `9aa63481` — DNSBL datapath swap (#3200)

`pfb_reload_unbound(..., $datapath = TRUE)` no longer falls back to `pfb_stop_start_unbound()` when the ADR-10 applied-generation handshake does not confirm in time. On every such reload the old failure mode was the reported outage: 30 s budget, `not confirmed in time -- falling back to Unbound restart`, DNS dropped on a ~1.1 M-entry box.

New semantics:

- The wait budget becomes `PFB_DNSBL_SWAP_WAIT` (1800 s, vs the old 30) — a self-unlock ceiling, not a failure verdict. The swap itself is unaffected by when PHP stops watching: the watcher applies the generation independently, so an early return costs nothing but caller contract (flush-skipping), which is correct while the new lists are not live.
- The wait now breaks immediately when the Unbound process dies (the reload-watcher thread lives inside it) — the one definite-broken case — and recovers via the restart path exactly as before. This is the practical failure latency: seconds, not the budget.
- A budget expiry with Unbound still running is **not** an error: log `swap still in flight, previous lists serving`, clear the `.sync` daemon-suppression marker, open the ADR-61 pending apply-ledger entry, and return FALSE. The existing tick reconciliation re-publishes the sentinel until the swap converges.
- The sentinel-flip-failure and RAM-gate restart branches are unchanged; `pfb_unbound_py_wait_control_applied`, the Python module, #3125 fingerprint restarts, and the #3198 Reports gate are untouched.

This covers every `$datapath = TRUE` caller (bulk `pfb_update_unbound`, cron/feed, and all four Alerts/Reports sites) — not just the whitelist path.

## Commit 2 — `f7731bbd` — firewall-configuration reload detached

Per owner directive, the #2878 bounded reload is superseded: pfSense's `/etc/rc.filter_configure_sync` is not ours to bound or kill, and the script signals filterd asynchronously — its exit never meant rules were live. `pfb_filter_reload_exec()` now:

- gates on `is_file()` + `is_executable()` — the one failure pfBlockerNG can own — logging and returning −1 so the pass is marked pending;
- otherwise fires the script detached (`> /dev/null 2>&1 < /dev/null &`): no capture pipe for the backgrounded child, no wait, no kill on expiry, and later apply stages no longer block on pfSense;
- `pfb_filter_reload_cmd()` is deleted (its only consumer was `exec()` itself).

## Test-first evidence

- New `PfbSyncStatusDnsblWritersTest` rows (timeout keeps resolver serving, watcher death recovers via restart) executed RED on untouched `origin/devel` (timeout still hit `pfb_stop_start_unbound`, log still contained `falling back to Unbound restart`), GREEN after with byte-identical test files. New `UnboundPyPublishTest` rows pin the liveness probe short-circuit and marker-first ordering.
- `FilterReloadBoundTest` rewritten for the detached contract; on the pre-change tree it failed 6/10 (30.0 s blocked wait, 124/-1 gating), passes 10/10 in ~0.5 s after.
- Independent verifier gate re-executed all red proofs on the committed test bytes (revert → FAIL, restore → PASS, blobs byte-identical), re-derived base parity for the suite's environmental failures, and ticked the full coverage matrix: **PASS**.

## Gate note

`run-gates.sh --diff origin/devel`: phpstan, phpcs, phplint, ruff (check + format), markdownlint, shell gates green. The full-suite phpunit leg shows 5 errors + a handful of failures that reproduce **identically on a pristine base scratch tree at the same commit** (SIGXFSZ status, `ValueError: Path must not be empty` at `pfblockerng.inc:9108`, Tick cadence wall-clock flakes, and `SourceScanningWiringPolicyTest`'s nested copy of the same base `Top1mDccDetectorTest` environment failure). None reference this diff; none fail on base parity.

## Follow-up (live verification)

The issue's risk-trigger row — a live datapath reload on a large DNSBL logging the new in-flight line (and *not* `Stopping Unbound Resolver`) — still wants one run on the reported box (10.1.41.3) or a smoke seat. The smoke module for the firewall seam was rewritten to the detached contract in this PR and awaits its live CI seat as usual.

## Summary by CodeRabbit

* **Improvements**
  * Firewall configuration reloads now start asynchronously, preventing slow reloads from blocking normal operations.
  * Launch failures are detected promptly, with affected changes retained for recovery.
  * DNSBL updates now wait longer for confirmation while continuing to serve the existing blocking configuration.
  * If the resolver remains active during an unconfirmed update, service continues without an unnecessary restart.
  * If the resolver stops responding, recovery automatically falls back to a restart path.

### CODE REVIEW

**Comment by andrebrait** [comment]
⏳ ***#3200***(***#3204***): verification status on head `f83f15d9`

**PR CI — green on the exact head.** Run [33969318890](https://github.com/pfBlockerNG/pfBlockerNG/actions/runs/33969318890) (`headSha=f83f15d9`), every check pass, including PHPUnit on PHP 8.3 / 8.4 / 8.5. The unit-suite failures seen on the authoring workstation were environmental (SIGXFSZ status, an empty-path `ValueError`, Tick cadence wall-clock flakes) and reproduce identically on a pristine base tree; CI's green PHPUnit legs confirm that reading.

**Targeted smoke — green on a live VM.** Run [33976083453](https://github.com/pfBlockerNG/pfBlockerNG/actions/runs/33976083453) (impacted scope, CE 2.8, `-k 'datapath_swap_budget_expiry or filter_reload'`): **4 passed, 1089 deselected**.

- `test_smoke_matrix.py::test_datapath_swap_budget_expiry_keeps_resolver_serving` — the #3200 invariant, proven on a real box: with the manifest corrupted so the watcher fail-closes, the real datapath fast path under a 1 s budget seam logged the in-flight line, cleared `.sync`, opened the ADR-61 apply entry, kept the previous snapshot serving, and left Unbound's pid unchanged. No `Stopping Unbound Resolver`, no restart fallback.
- `test_filter_reload_healthy_pass_reaches_finite_terminal_state`, `test_filter_reload_stall_does_not_block_the_pass`, `test_filter_reload_launch_gate_marks_pending_and_names_itself` — the detached filter-reload contract, including a stalled reload no longer blocking the pass.

**Full smoke fan-out — dispatched.** The narrowed run does not cover the untouched matrix rows against the changed production paths (every datapath reload and every filter-configure pass), so a `scope=full` fan-out over all ci:true legs is now running; its verdict is the remaining gate before landing.

**Comment by BBcan177** [comment]
Live verification on the reported box ( pfSense Plus 25.11, DNSBL python, ~1.1M loaded entries).

**Package:** nightly installer only (`fetch …/install.sh | sh -s -- --channel nightly`) — `58f7b57` → `20260905123014.713025f`. No pfSense `pkg upgrade`. NUT/Snort/OpenVPN-export/`pfSense-repoc` left unchanged. `pfb_unbound.py` bytes unchanged; Unbound pid `93154` (started 11:16:10) survived the package enable pass.

**Overlay:** PR head `f83f15d9` `pfblockerng.inc` + `pfblockerng_apply.inc` onto that nightly. Backups: `*.bak.202609051206`.

**Datapath probe** (`pfb_reload_unbound('enabled', FALSE, FALSE, TRUE)` at 12:07:12):

```text
PROBE_START ts=12:07:12 pid=93154 applied=109 sentinel=109
# t=2s  sentinel=110 applied=109  dns=216.239.38.120
# t=30s sentinel=110 applied=109  dns=216.239.38.120   ← old 30s budget; no restart
# t=40s sentinel=110 applied=110  dns=216.239.38.120
PROBE_END ok=true elapsed=40.563 pid=93154 applied=110 sentinel=110
```

```text
2026-09-05 12:07:12 Reloading Unbound Resolver (DNSBL python) [ zero-downtime swap ] completed
```

No `falling back to Unbound restart`, no `Stopping Unbound Resolver`. `.sync` cleared.

**Hourly cron** (the ticket's every-:15 outage):

```text
2026-09-05 12:15:27  UPDATE PROCESS START [ v20260905123014.713025f ]
2026-09-05 12:15:50 Reloading Unbound Resolver (DNSBL python) [ zero-downtime swap ] completed
2026-09-05 12:16:31 DNSBL update [ feed lines: 1224842 | loaded entries: 1102851 ]... completed
2026-09-05 12:17:02  UPDATE PROCESS ENDED
```

Unbound still pid `93154` at 12:18:22. DNS still answering. Filter-reload detach was overlaid, not separately exercised (no filter-configure this pass: alias updates only).

**Comment by andrebrait** [comment]
🚧 ***#3200***(***#3204***): full smoke fan-out verdict — red, but red identically on `devel`

Run [33976401293](https://github.com/pfBlockerNG/pfBlockerNG/actions/runs/33976401293) (`scope=full`, all ci:true legs) failed its AND gate. The failure set is `devel`'s own pre-existing smoke red, tracked as #3041 ([nightly-red] since 09-03; every `devel` fan-out from 09-03 onward is `failure`). Compared leg-for-leg and shard-for-shard against `devel`'s latest nightly [33962709948](https://github.com/pfBlockerNG/pfBlockerNG/actions/runs/33962709948):

| Shard (every leg) | This branch | devel nightly | Verdict |
| --- | --- | --- | --- |
| shard 1/3 | 3 failed, **102 passed**, 6 errors | 3 failed, 101 passed, 6 errors | same names; +1 pass = the new #3200 row |
| shard 3/3 (CE 2.8) | **1 failed**, 129 passed | 2 failed, 128 passed | same `recompute_v6` failure; one devel failure does not reproduce here |

Identical names on both sides:

- shard 1 — `test_smoke_feeds.py::test_extraction_refuses_archive_supplied_metadata`, `::test_zip_extraction_failure_rejected_not_empty`, `::test_geoip_partway_extraction_keeps_the_published_tree_byte_identical`; 6 errors in `test_smoke_adr40.py`.
- shard 3 — `test_smoke_ip_recompute.py::test_recompute_v6_snapshot_round_trips_across_static_pass`, failing on CE 2.8, CE 2.9, Plus 26.03 and Plus 26.07 in both runs.

So no fan-out failure traces to this diff, and the two datapath/filter surfaces it changes are covered green by the targeted run above (4/4 live). The remaining decision is a landing one: either #3041 is fixed first so a genuinely green fan-out can gate this PR, or this PR lands on the documented parity plus the green targeted run. Flagging rather than choosing.

**Comment by andrebrait** [comment]
👀 ***#3200***(***#3204***): adversarial review record — condensed 4-in-1 round

**Leg:** single condensed round covering all four lenses (contract conformance, correctness + hostile inputs, test honesty, over-engineering), per the active review experiment.
**Head SHA reviewed:** `f83f15d9`
**Harness / model / lever:** OMP subagent (`reviewer` role) · `anthropic/claude-sonnet-5` · thinking-level high (the runtime lever is not introspectable from inside the subagent; recorded as configured).
**Verdict:** no blocking defect survived scrutiny. Four production mutations were executed in a private scratch copy to confirm the new tests fail on regression; two further mutations exposed real coverage gaps.

**Findings ledger**

1. Liveness granularity: process-liveness is not thread-liveness — a dead watcher *thread* inside a live Unbound would not trip the probe. Investigated hard, not reproducible against the module's current defensive coding; the inaccuracy that remains is a doc claim stating process==thread liveness as fact. → to fix: qualify the claim in `architecture-notes.md`.
2. `FilterReloadBoundTest` coverage gap A — the `?? '/etc/rc.filter_configure_sync'` fallback default is unit-test-dark. Mutation to a typo'd path left 10/10 green. → to fix: add a row.
3. `FilterReloadBoundTest` coverage gap B — `< /dev/null` stdin redirection unguarded. Removing it left 10/10 green. → to fix: add a row.
4. Copilot inline (`pfblockerng.inc:12206`): `pfb_unbound_py_gc()` takes the ~0.25 s publication lock even when convergence is already known false. → to fix: guard the call.
5. Copilot inline (`pfblockerng.inc:12228`): the new recovery log line is logtype 2 with no newline framing, reproducing the `error.log` concatenation anti-pattern this issue's own body named — and it is a brand-new line, so writing it cleanly costs nothing. → to fix.
6. `_open_apply_ledger_count()` in the new smoke row reimplements `helpers._php_read_scalar()`. → to fix: reuse the helper.
7. `pfb_unbound_py_wait_applied()`'s `$timeout_s = 30` default is now vestigial (every caller passes explicitly) and its doc comment describes a budget nothing uses. → to fix.
8. `PFB_DNSBL_SWAP_WAIT` value-anchor comment (why 1800 matches `PFB_REENTRY_TIMEOUT` precedent without deriving from it). → to fix.
9. Residual structural duplication between the stall and launch-gate smoke rows. → skipped: this repo documents a deliberate bias toward some test-fragment independence (`confirmRecorderCmd()`'s docblock cites a real incident where shared fragments let a typo certify an unshared mechanism).
10. `$GLOBALS['pfb_test_swap_wait_s']` `(int)` cast edge cases (array → 1, overflow). → skipped: grep shows zero production writers; the global is reachable only from test code, so the hostile framing does not apply.

**Refuted and recorded** (probes executed, findings dropped): detached `exec()` capture-pipe/stdin/quoting and hostile script paths (non-blocking in 0.0036 s; `escapeshellarg` proven safe against `; $(whoami) '`; every bad path gated to −1 before exec); a new ordering race against the filter-daemon stage (the old design never guaranteed live rules either — the reload signals filterd asynchronously); `?callable $alive` as speculative flexibility (the shared primitive genuinely has two callers with different liveness semantics); the test seam as a novel pattern (three sibling `$GLOBALS['pfb_test_*']` seams pre-date it); `PFB_DNSBL_SWAP_WAIT` as duplicating the #2851 operator knob (an in-process poll bound versus an operator-configurable external-subprocess kill bound — coupling them would let an unrelated admin change alter swap behavior).

Fixes for items 1–8 follow in one batched commit; each implements its reviewer's concrete suggestion. Copilot's two threads get replies citing that commit.

File: src/usr/local/pkg/pfblockerng/pfblockerng.inc [hunk 1]
```diff
@@ -12182,18 +12206,26 @@ function pfb_reload_unbound($mode, $cache=FALSE, $pfbpython=FALSE, $datapath=FAL
 					return TRUE;
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply src/usr/local/pkg/pfblockerng/pfblockerng.inc:12218]: Applied in `ddf03eab`. The call is now guarded so the publication lock is only taken when convergence is real:

```php
$converged = $apply_ledger();
if ($converged) {
    pfb_unbound_py_gc($converged);
}
```

Guarded rather than removed on purpose: a swap that lands between the wait expiring and this ledger read still reports `TRUE`, and that case genuinely has generations to collect — so the useful path keeps its GC while the guaranteed no-op path (every in-flight expiry) skips the ~0.25 s lock.

File: src/usr/local/pkg/pfblockerng/pfblockerng.inc [hunk 2]
```diff
@@ -12182,18 +12206,26 @@ function pfb_reload_unbound($mode, $cache=FALSE, $pfbpython=FALSE, $datapath=FAL
...
+					// to write it, and tick reconciliation keys on this entry.
+					$converged = $apply_ledger();
+					pfb_unbound_py_gc($converged);
+					return FALSE;
+				}
+
+				// Unbound exited during the swap: nothing will ever apply the generation,
+				// so fall through to the restart tail -- restarting is what brings DNS
+				// back with the new manifest.
+				pfb_logger(" swap unconfirmed: Unbound exited during the swap -- recovering via restart", 2);
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply src/usr/local/pkg/pfblockerng/pfblockerng.inc]: Applied in `ddf03eab`. The line is now framed like its sibling fail-safe branch instead of continuing the previous one:

```php
pfb_logger("\nADR-10: swap unconfirmed -- Unbound exited during the swap; recovering via restart", 2);
```

You were right that this one was free: the issue body's "do not mix in a logging fix" applied to the pre-existing `not confirmed in time` line, not to a line this PR introduces. The leading newline is what stops the `error.log` concatenation, and it matches the adjacent `\nADR-10: sentinel flip failed ...` logtype-2 message exactly. The in-flight line one branch above stays a logtype-1 continuation of the `[ zero-downtime swap ]` line — it never reaches `error.log`, so it has no concatenation to cause.

**Comment by andrebrait** [comment]
@coderabbitai review

File: tests/php/PfbSyncStatusDnsblWritersTest.php [hunk 3]
```diff
@@ -82,7 +82,7 @@ protected function tearDown(): void
 			}
 		}
 		$this->savedG = [];
-		unset($GLOBALS['pfb_test_process_running'], $GLOBALS['config']['unbound']);
+		unset($GLOBALS['pfb_test_process_running'], $GLOBALS['pfb_test_swap_wait_s'], $GLOBALS['config']['unbound']);
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/php/PfbSyncStatusDnsblWritersTest.php]: Applied in `d75efd75`. One correction to the location: the unset is in `tearDown()`, not `setUp()` — but the defect is exactly as described, and it was introduced by this PR (`git blame` → `9aa63481`), so it is in scope.

`setUp()` now saves the key the same way the file already saves `$GLOBALS['pfb']` and `$GLOBALS['g']` (sentinel `false` = was unset), and `tearDown()` restores instead of deleting:

```php
foreach (['unbound'] as $k) {
    $this->savedConfig[$k] = array_key_exists($k, $GLOBALS['config'] ?? []) ? $GLOBALS['config'][$k] : false;
}
```

So a bootstrap-owned `config['unbound']` survives this class, and the order-dependence is gone.

File: tests/smoke/test_smoke_matrix.py [hunk 4]
```diff
@@ -1053,6 +1053,134 @@ def _pyerr_logged(_a: h.DnsAnswer) -> bool:
...
+    return out[start + len(opening) : end]
+
+
+def _open_apply_ledger_count(vm: SmokeVM) -> int:
+    """Count open (dnsbl) ADR-61 sync-status entries via the shared delimited-read helper."""
+    return int(
+        h._php_read_scalar(  # noqa: SLF001 -- suite-internal helper, used the same way elsewhere here
+            vm,
+            "require_once('/usr/local/pkg/pfblockerng/pfblockerng.inc');\npfb_global();",
+            "count(pfb_sync_status_list_open($GLOBALS['pfb']['dbdir'], 'dnsbl'))",
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/smoke/test_smoke_matrix.py]: Applied in `d75efd75`. You were right that the group count could be satisfied by the wrong record — the helper now filters to the exact key and the row asserts a closed baseline first:

```php
count(array_filter(pfb_sync_status_list_open($GLOBALS['pfb']['dbdir'], 'dnsbl'),
    static fn($e) => ($e['item'] ?? '') === 'dnsbl' && ($e['stage'] ?? '') === 'apply'))
```

```python
assert ledger_before == 0, "baseline: the (dnsbl, dnsbl, apply) entry must start CLOSED ..."
assert _open_apply_ledger_count(deployed_vm) >= 1, "the unconfirmed swap must OPEN ..."
```

Without the closed baseline the delta assertion could still have been satisfied by pre-existing state, so both halves were needed.

**Comment by andrebrait** [comment]
👀 ***#3200***(***#3204***): findings ledger — every item with its outcome

**Round 1 — condensed 4-in-1 adversarial review** (head reviewed `f83f15d9`)

1. Liveness granularity: a process probe cannot see a dead watcher *thread* — `fixed@ddf03eab` (the architecture note now states the limitation instead of claiming process == thread liveness).
2. `FilterReloadBoundTest` gap: absent-injection fallback path unit-test-dark — `fixed@ddf03eab`. The first attempt failed its own mutation test (a bare substring let `/etc/rc.filter_configure_sync_TYPO` pass); the assertion is now bracket-delimited and the drifted-default mutant dies.
3. `FilterReloadBoundTest` gap: `< /dev/null` stdin redirection — `partially fixed@ddf03eab`, remainder `skipped: evidence`. A new row pins that the child never reads the launcher's stdin and kills a synchronous-launch mutant; the redirect *alone* is untestable because POSIX assigns `/dev/null` to an async list's stdin before explicit redirection (`printf PAYLOAD | sh -c 'read l && echo got || echo closed &'` → `closed`; same command in the foreground → `got`). It stays as documented intent.
4. `pfb_unbound_py_gc()` took the publication lock on a known-false convergence — `fixed@ddf03eab` (guarded, not removed: a swap landing between the wait and the ledger read still collects).
5. New watcher-death log line had no newline framing — `fixed@ddf03eab` (framed like its sibling fail-safe line, so `error.log` cannot concatenate).
6. `_open_apply_ledger_count()` reimplemented an existing helper — `fixed@ddf03eab`.
7. Vestigial `$timeout_s = 30` default — `fixed@ddf03eab` (now `PFB_DNSBL_SWAP_WAIT`; golden signature snapshot refreshed in `6bde7741`).
8. Missing rationale for the 1800 value — `fixed@ddf03eab`.
9. Residual duplication between two smoke rows — `skipped: evidence`. This repo documents a deliberate bias toward test-fragment independence (`confirmRecorderCmd()`'s docblock cites an incident where shared fragments let a typo certify an unshared mechanism).
10. Test-seam `(int)` cast edge cases — `skipped: evidence` (grep shows zero production writers; the global is reachable only from test code).

**Round 2 — CodeRabbit** (head reviewed `6bde7741`, Merge Risk 🔵 Low)

11. `config['unbound']` deleted instead of restored, making later tests order-dependent — `fixed@d75efd75`.
12. Ledger count included download records, so the apply assertion could pass without the apply entry opening — `fixed@d75efd75` (exact-key filter plus a closed-baseline assertion).
13. `_swap_reload_script()` could leave a truncated appliance reload script on a session-scoped VM if `tee`/`chmod` failed after the backup existed — `fixed@d75efd75` (the helper restores on any failure).
14. Non-recursive PHPUnit temp cleanup leaked a nested directory — `fixed@d75efd75`.
15. Liveness probe forked `pgrep` on every 200 ms poll — up to ~9000 forks across a full budget, beside the RAM-heavy rebuild — `fixed@d75efd75`, re-applied `4ebe4a57`. Best catch of the round.

**Two process corrections worth recording**

- Item 15's production change was briefly pushed *without* its throttle: a mutation run restored the file from the last commit while the throttle was still uncommitted, so `d75efd75` carried the guard alone. Caught by the full-suite failing-set diff (not by verdict comparison), re-applied in `4ebe4a57`, and the guard is now rate-based (cap derived from measured elapsed time) instead of a fixed count that passed in isolation and failed under load.
- Earlier in this PR, gate verdicts were compared instead of failing sets, which let a stale `function_inventory` golden hide inside the known environmental failures until CI named it. The diff-the-failing-set discipline is what caught both.

**Re-review status:** every APPLY above implements its reviewer's own concrete suggestion (tests adjusted where the suggestion touched tests), which is landing.md's stated exemption from another leg round. No second review-bot ask: its own diffs were applied, and asking again for that spends a slot for nothing.

Exact-head CI is green on `4ebe4a57` (28 checks; the two `skipping` rows are conditional legs).

**Comment by andrebrait** [comment]
🏁 ***#3200***(***#3204***): landed on `devel`

**Path:** maintainer-local signed fast-forward (`cab1cc81..389db199`), preserving all seven signed commits — GitHub merge commits and server-side rebase merges stay disabled.

**Base fence:** the first attempt stopped — `devel` had advanced to `cab1cc81` (`docs: record licensed ASN/GeoIP smoke skip boundary`, touching only `.agents/context/smoke.md`, zero overlap with this diff's eleven files), so the review verdicts stood and the rebase was mechanical. Rebased, graph re-verified fresh, force-with-lease pushed, exact-head CI re-run, fence re-checked (`remote_pr_head == reviewed_sha == 389db199`, base `cab1cc81`), then fast-forwarded. `git merge-base --is-ancestor` confirms the landed SHA.

**Signatures:** all seven commits report `G` (good) under the configured identity.

**Exact-head CI:** green on `389db199` (28 checks; the two `skipping` rows are conditional legs — Benchmark kill-gates and the UI fan-out).

**Review:** one condensed 4-in-1 adversarial round (`f83f15d9`), Copilot (`f7731bbd`), CodeRabbit (`6bde7741`, Merge Risk 🔵 Low). Fifteen findings, every one with a recorded outcome in the ledger above: twelve fixed, three skipped with executed evidence. Two process corrections are recorded there too — a production change briefly pushed without its throttle (caught by a failing-set diff, not a verdict comparison), and a stale golden fixture that hid inside the known environmental failures until CI named it.

**Live verification:** the targeted smoke run proved the new `#3200` invariant on a real CE 2.8 VM (`test_datapath_swap_budget_expiry_keeps_resolver_serving` plus the three rewritten firewall rows: 4 passed). The full fan-out is red identically on `devel` — the pre-existing `#3041` baseline (feeds-extraction trio, adr40 sextet, `recompute_v6` on every leg), verified shard-for-shard against `devel`'s own nightly. Owner confirmed the fix works on the reporting appliance.
